File: server/node/runtime/native/src/connection.c
/*
 * Copyright © 2008 Kristian Høgsberg
 * Copyright © 2013 Jason Ekstrand
 * Copyright © 2018 Erik De Rijcke
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice (including the
 * next paragraph) shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT.  IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#define _GNU_SOURCE

#include <math.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <stdio.h>
#include <errno.h>
#include <sys/uio.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <time.h>

#include "connection.h"
#include "wayland-util.h"
#include "wayland-private.h"
#include "wayland-os.h"

struct wl_buffer {
    char data[4096];
    uint32_t head, tail;
};

#define MASK(i) ((i) & 4095)

#define MAX_FDS_OUT    28
#define CLEN        (CMSG_LEN(MAX_FDS_OUT * sizeof(int32_t)))

struct wl_connection {
    struct wl_buffer in, out;
    struct wl_buffer fds_in, fds_out;
    int fd;
    int want_flush;
};

static int
wl_buffer_put(struct wl_buffer *b, const void *data, size_t count) {
    uint32_t head, size;

    if (count > sizeof(b->data)) {
        wl_log("Data too big for buffer (%d > %d).\n",
               count, sizeof(b->data));
        errno = E2BIG;
        return -1;
    }

    head = MASK(b->head);
    if (head + count <= sizeof b->data) {
        memcpy(b->data + head, data, count);
    } else {
        size = sizeof b->data - head;
        memcpy(b->data + head, data, size);
        memcpy(b->data, (const char *) data + size, count - size);
    }

    b->head += count;

    return 0;
}

static void
wl_buffer_put_iov(struct wl_buffer *b, struct iovec *iov, int *count) {
    uint32_t head, tail;

    head = MASK(b->head);
    tail = MASK(b->tail);
    if (head < tail) {
        iov[0].iov_base = b->data + head;
        iov[0].iov_len = tail - head;
        *count = 1;
    } else if (tail == 0) {
        iov[0].iov_base = b->data + head;
        iov[0].iov_len = sizeof b->data - head;
        *count = 1;
    } else {
        iov[0].iov_base = b->data + head;
        iov[0].iov_len = sizeof b->data - head;
        iov[1].iov_base = b->data;
        iov[1].iov_len = tail;
        *count = 2;
    }
}

static void
wl_buffer_get_iov(struct wl_buffer *b, struct iovec *iov, int *count) {
    uint32_t head, tail;

    head = MASK(b->head);
    tail = MASK(b->tail);
    if (tail < head) {
        iov[0].iov_base = b->data + tail;
        iov[0].iov_len = head - tail;
        *count = 1;
    } else if (head == 0) {
        iov[0].iov_base = b->data + tail;
        iov[0].iov_len = sizeof b->data - tail;
        *count = 1;
    } else {
        iov[0].iov_base = b->data + tail;
        iov[0].iov_len = sizeof b->data - tail;
        iov[1].iov_base = b->data;
        iov[1].iov_len = head;
        *count = 2;
    }
}

static void
wl_buffer_copy(struct wl_buffer *b, void *data, size_t count) {
    uint32_t tail, size;

    tail = MASK(b->tail);
    if (tail + count <= sizeof b->data) {
        memcpy(data, b->data + tail, count);
    } else {
        size = sizeof b->data - tail;
        memcpy(data, b->data + tail, size);
        memcpy((char *) data + size, b->data, count - size);
    }
}

static uint32_t
wl_buffer_size(struct wl_buffer *b) {
    return b->head - b->tail;
}

struct wl_connection *
wl_connection_create(int fd) {
    struct wl_connection *connection;

    connection = zalloc(sizeof *connection);
    if (connection == NULL)
        return NULL;

    connection->fd = fd;

    return connection;
}

static void
close_fds(struct wl_buffer *buffer, int max) {
    int32_t fds[sizeof(buffer->data) / sizeof(int32_t)], i, count;
    size_t size;

    size = wl_buffer_size(buffer);
    if (size == 0)
        return;

    wl_buffer_copy(buffer, fds, size);
    count = size / sizeof fds[0];
    if (max > 0 && max < count)
        count = max;
    size = count * sizeof fds[0];
    for (i = 0; i < count; i++)
        close(fds[i]);
    buffer->tail += size;
}

int
wl_connection_destroy(struct wl_connection *connection) {
    int fd = connection->fd;

    close_fds(&connection->fds_out, -1);
    close_fds(&connection->fds_in, -1);
    free(connection);

    return fd;
}

void
wl_connection_copy(struct wl_connection *connection, void *data, size_t size) {
    wl_buffer_copy(&connection->in, data, size);
}

static void
build_cmsg(struct wl_buffer *buffer, char *data, int *clen) {
    struct cmsghdr *cmsg;
    size_t size;

    size = wl_buffer_size(buffer);
    if (size > MAX_FDS_OUT * sizeof(int32_t))
        size = MAX_FDS_OUT * sizeof(int32_t);

    if (size > 0) {
        cmsg = (struct cmsghdr *) data;
        cmsg->cmsg_level = SOL_SOCKET;
        cmsg->cmsg_type = SCM_RIGHTS;
        cmsg->cmsg_len = CMSG_LEN(size);
        wl_buffer_copy(buffer, CMSG_DATA(cmsg), size);
        *clen = cmsg->cmsg_len;
    } else {
        *clen = 0;
    }
}

static int
decode_cmsg(struct wl_buffer *buffer, struct msghdr *msg) {
    struct cmsghdr *cmsg;
    size_t size, max, i;
    int overflow = 0;

    for (cmsg = CMSG_FIRSTHDR(msg); cmsg != NULL;
         cmsg = CMSG_NXTHDR(msg, cmsg)) {
        if (cmsg->cmsg_level != SOL_SOCKET ||
            cmsg->cmsg_type != SCM_RIGHTS)
            continue;

        size = cmsg->cmsg_len - CMSG_LEN(0);
        max = sizeof(buffer->data) - wl_buffer_size(buffer);
        if (size > max || overflow) {
            overflow = 1;
            size /= sizeof(int32_t);
            for (i = 0; i < size; i++)
                close(((int *) CMSG_DATA(cmsg))[i]);
        } else if (wl_buffer_put(buffer, CMSG_DATA(cmsg), size) < 0) {
            return -1;
        }
    }

    if (overflow) {
        errno = EOVERFLOW;
        return -1;
    }

    return 0;
}

int
wl_connection_flush(struct wl_connection *connection) {
    struct iovec iov[2];
    struct msghdr msg;
    char cmsg[CLEN];
    int len = 0, count, clen;
    uint32_t tail;

    if (!connection->want_flush)
        return 0;

    tail = connection->out.tail;
    while (connection->out.head - connection->out.tail > 0) {
        wl_buffer_get_iov(&connection->out, iov, &count);

        build_cmsg(&connection->fds_out, cmsg, &clen);

        msg.msg_name = NULL;
        msg.msg_namelen = 0;
        msg.msg_iov = iov;
        msg.msg_iovlen = count;
        msg.msg_control = (clen > 0) ? cmsg : NULL;
        msg.msg_controllen = clen;
        msg.msg_flags = 0;

        do {
            len = sendmsg(connection->fd, &msg,
                          MSG_NOSIGNAL | MSG_DONTWAIT);
        } while (len == -1 && errno == EINTR);

        if (len == -1)
            return -1;

        close_fds(&connection->fds_out, MAX_FDS_OUT);

        connection->out.tail += len;
    }

    connection->want_flush = 0;

    return connection->out.head - tail;
}

uint32_t
wl_connection_pending_input(struct wl_connection *connection) {
    return wl_buffer_size(&connection->in);
}

int
wl_connection_read(struct wl_connection *connection) {
    struct iovec iov[2];
    struct msghdr msg;
    char cmsg[CLEN];
    int len, count, ret;

    if (wl_buffer_size(&connection->in) >= sizeof(connection->in.data)) {
        errno = EOVERFLOW;
        return -1;
    }

    wl_buffer_put_iov(&connection->in, iov, &count);

    msg.msg_name = NULL;
    msg.msg_namelen = 0;
    msg.msg_iov = iov;
    msg.msg_iovlen = count;
    msg.msg_control = cmsg;
    msg.msg_controllen = sizeof cmsg;
    msg.msg_flags = 0;

    do {
        len = wl_os_recvmsg_cloexec(connection->fd, &msg, MSG_DONTWAIT);
    } while (len < 0 && errno == EINTR);

    if (len <= 0)
        return len;

    ret = decode_cmsg(&connection->fds_in, &msg);
    if (ret)
        return -1;

    connection->in.head += len;

    return wl_connection_pending_input(connection);
}

int
wl_connection_write(struct wl_connection *connection,
                    const void *data, size_t count) {
    if (connection->out.head - connection->out.tail +
        count > ARRAY_LENGTH(connection->out.data)) {
        connection->want_flush = 1;
        if (wl_connection_flush(connection) < 0)
            return -1;
    }

    if (wl_buffer_put(&connection->out, data, count) < 0)
        return -1;

    connection->want_flush = 1;

    return 0;
}

int
wl_connection_get_fd(struct wl_connection *connection) {
    return connection->fd;
}

static int
wl_connection_put_fd(struct wl_connection *connection, int32_t fd) {
    if (wl_buffer_size(&connection->fds_out) == MAX_FDS_OUT * sizeof fd) {
        connection->want_flush = 1;
        if (wl_connection_flush(connection) < 0)
            return -1;
    }

    return wl_buffer_put(&connection->fds_out, &fd, sizeof fd);
}

size_t
wl_connection_fds_in_count(struct wl_connection *connection) {
    size_t count = 0;
    int tail = connection->fds_in.tail;
    while (tail != connection->fds_in.head) {
        count++;
        tail += sizeof(int);
    }
    return count;
}


int
wl_connection_read_next_fd_in(struct wl_connection *connection, int* fd) {
    if (connection->fds_in.tail == connection->fds_in.head) {
        return -1;
    }

    wl_buffer_copy(&connection->fds_in, fd, sizeof(int));
    connection->fds_in.tail += sizeof(int);
    return 0;
}

File: common/src/Connection.ts
/*
MIT License

Copyright (c) 2020 Erik De Rijcke

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
const textDecoder = new TextDecoder('utf8');

export class WlObject {
  readonly id: number;
  // @ts-ignore
  private _destroyResolver: () => void;
  private readonly _destroyPromise: Promise<void> = new Promise(resolve => this._destroyResolver = resolve)
  private _destroyListeners: ((wlObject: WlObject) => void)[] = [];

  constructor(id: number) {
    this.id = id
    this._destroyPromise.then(() => this._destroyListeners.forEach(destroyListener => destroyListener(this)))
  }

  destroy() {
    this._destroyResolver()
  }

  addDestroyListener(destroyListener: (wlObject: WlObject) => void) {
    this._destroyListeners.push(destroyListener)
  }

  removeDestroyListener(destroyListener: (wlObject: WlObject) => void) {
    this._destroyListeners = this._destroyListeners.filter((item) => item !== destroyListener)
  }

  onDestroy() {
    return this._destroyPromise
  }
}

export class Fixed {
  readonly _raw: number;

  static parse(data: number): Fixed {
    return new Fixed((data * 256.0) >> 0)
  }

  /**
   * Represent fixed as a signed 24-bit integer.
   *
   */
  asInt(): number {
    return ((this._raw / 256.0) >> 0)
  }

  /**
   * Represent fixed as a signed 24-bit number with an 8-bit fractional part.
   *
   */
  asDouble(): number {
    return this._raw / 256.0
  }

  /**
   * use parseFixed instead
   * @param {number}raw
   */
  constructor(raw: number) {
    this._raw = raw
  }
}

export class WebFD {
  readonly fd: number;
  readonly type: "ImageBitmap" | "ArrayBuffer" | "MessagePort|OffscreenCanvas";
  readonly url: URL;
  private readonly _onGetTransferable: (webFd: WebFD) => Promise<Transferable>;
  private readonly _onClose: (webFd: WebFD) => void;

  constructor(
    fd: number,
    fdType: 'ImageBitmap' | 'ArrayBuffer' | 'MessagePort|OffscreenCanvas',
    fdURL: URL,
    onGetTransferable: (webFd: WebFD) => Promise<Transferable>,
    onClose: (webFd: WebFD) => void
  ) {
    this.fd = fd
    this.type = fdType
    this.url = fdURL
    this._onGetTransferable = onGetTransferable
    this._onClose = onClose
  }

  async getTransferable(): Promise<Transferable> {
    return await this._onGetTransferable(this)
  }

  close() {
    this._onClose(this)
  }
}

export interface MessageMarshallingContext<V extends number | WebFD | Fixed | WlObject | 0 | string | ArrayBufferView,
  T extends 'u' | 'h' | 'i' | 'f' | 'o' | 'n' | 's' | 'a',
  S extends 0 | 4 | number> {
  readonly value: V,
  readonly type: T,
  readonly size: number,
  readonly optional: boolean,
  readonly _marshallArg: (wireMsg: { buffer: ArrayBuffer, fds: Array<WebFD>, bufferOffset: number }) => void
}

export interface WlMessage {
  buffer: Uint32Array,
  fds: Array<WebFD>,
  bufferOffset: number,
  consumed: number,
  size: number
}

export interface SendMessage {
  buffer: ArrayBuffer,
  fds: Array<WebFD>
}

export function uint(arg: number): MessageMarshallingContext<number, 'u', 4> {
  return {
    value: arg,
    type: 'u',
    size: 4,
    optional: false,
    _marshallArg: function (wireMsg) {
      new Uint32Array(wireMsg.buffer, wireMsg.bufferOffset, 1)[0] = arg
      wireMsg.bufferOffset += this.size
    }
  }
}

export function uintOptional(arg: number): MessageMarshallingContext<number, 'u', 4> {
  return {
    value: arg,
    type: 'u',
    size: 4,
    optional: true,
    _marshallArg: function (wireMsg) {
      new Uint32Array(wireMsg.buffer, wireMsg.bufferOffset, 1)[0] = (arg === null ? 0 : arg)
      wireMsg.bufferOffset += this.size
    }
  }
}

export function fileDescriptor(arg: WebFD): MessageMarshallingContext<WebFD, 'h', 0> {
  return {
    value: arg,
    type: 'h',
    size: 0, // file descriptors are not added to the message size because they are somewhat considered meta data.
    optional: false,
    _marshallArg: function (wireMsg) {
      wireMsg.fds.push(arg)
    }
  }
}

export function fileDescriptorOptional(arg: WebFD): MessageMarshallingContext<WebFD, 'h', 0> {
  return {
    value: arg,
    type: 'h',
    size: 0, // file descriptors are not added to the message size because they are not part of the unix socket message buffer.
    optional: true,
    _marshallArg: function (wireMsg) {
      wireMsg.fds.push(this.value)
    }
  }
}

export function int(arg: number): MessageMarshallingContext<number, 'i', 4> {
  return {
    value: arg,
    type: 'i',
    size: 4,
    optional: false,
    _marshallArg: function (wireMsg) {
      new Int32Array(wireMsg.buffer, wireMsg.bufferOffset, 1)[0] = this.value
      wireMsg.bufferOffset += this.size
    }
  }
}

export function intOptional(arg: number): MessageMarshallingContext<number, 'i', 4> {
  return {
    value: arg,
    type: 'i',
    size: 4,
    optional: true,
    _marshallArg: function (wireMsg) {
      new Int32Array(wireMsg.buffer, wireMsg.bufferOffset, 1)[0] = (arg === null ? 0 : this.value)
      wireMsg.bufferOffset += this.size
    }
  }
}

export function fixed(arg: Fixed): MessageMarshallingContext<Fixed, 'f', 4> {
  return {
    value: arg,
    type: 'f',
    size: 4,
    optional: false,
    _marshallArg: function (wireMsg) {
      new Int32Array(wireMsg.buffer, wireMsg.bufferOffset, 1)[0] = this.value._raw
      wireMsg.bufferOffset += this.size
    }
  }
}

export function fixedOptional(arg: Fixed): MessageMarshallingContext<Fixed, 'f', 4> {
  return {
    value: arg,
    type: 'f',
    size: 4,
    optional: true,
    _marshallArg: function (wireMsg) {
      new Int32Array(wireMsg.buffer, wireMsg.bufferOffset, 1)[0] = (arg === null ? 0 : this.value._raw)
      wireMsg.bufferOffset += this.size
    }
  }
}

export function object(arg: WlObject): MessageMarshallingContext<WlObject, 'o', 4> {
  return {
    value: arg,
    type: 'o',
    size: 4,
    optional: false,
    /**
     *
     * @param {{buffer: ArrayBuffer, fds: Array<WebFD>, bufferOffset: number}} wireMsg
     * @private
     */
    _marshallArg: function (wireMsg) {
      new Uint32Array(wireMsg.buffer, wireMsg.bufferOffset, 1)[0] = this.value.id
      wireMsg.bufferOffset += this.size
    }
  }
}

export function objectOptional(arg: WlObject): MessageMarshallingContext<WlObject, 'o', 4> {
  return {
    value: arg,
    type: 'o',
    size: 4,
    optional: true,
    /**
     *
     * @param {{buffer: ArrayBuffer, fds: Array<WebFD>, bufferOffset: number}} wireMsg
     * @private
     */
    _marshallArg: function (wireMsg) {
      new Uint32Array(wireMsg.buffer, wireMsg.bufferOffset, 1)[0] = (arg === null ? 0 : this.value.id)
      wireMsg.bufferOffset += this.size
    }
  }
}

export function newObject(): MessageMarshallingContext<0, 'n', 4> {
  return {
    value: 0, // id filled in by _marshallConstructor
    type: 'n',
    size: 4,
    optional: false,
    /**
     * @param {{buffer: ArrayBuffer, fds: Array<WebFD>, bufferOffset: number}} wireMsg
     * @private
     */
    _marshallArg: function (wireMsg) {
      new Uint32Array(wireMsg.buffer, wireMsg.bufferOffset, 1)[0] = this.value
      wireMsg.bufferOffset += this.size
    }
  }
}

export function string(arg: string): MessageMarshallingContext<string, 's', number> {
  return {
    value: `${arg}\0`,
    type: 's',
    size: 4 + (function () {
      // fancy logic to calculate size with padding to a multiple of 4 bytes (int).
      // length+1 for null terminator
      return (arg.length + 1 + 3) & ~3
    })(),
    optional: false,
    _marshallArg: function (wireMsg) {
      new Uint32Array(wireMsg.buffer, wireMsg.bufferOffset, 1)[0] = this.value.length

      const strLen = this.value.length
      const buf8 = new Uint8Array(wireMsg.buffer, wireMsg.bufferOffset + 4, strLen)
      for (let i = 0; i < strLen; i++) {
        buf8[i] = this.value[i].charCodeAt(0)
      }
      wireMsg.bufferOffset += this.size
    }
  }
}

export function stringOptional(arg: string): MessageMarshallingContext<string, 's', number> {
  return {
    value: `${arg}\0`,
    type: 's',
    size: 4 + (function () {
      if (arg === null) {
        return 0
      } else {
        // fancy logic to calculate size with padding to a multiple of 4 bytes (int).
        // length+1 for null terminator
        return (arg.length + 1 + 3) & ~3
      }
    })(),
    optional: true,
    _marshallArg: function (wireMsg) {
      if (this.value === null) {
        new Uint32Array(wireMsg.buffer, wireMsg.bufferOffset, 1)[0] = 0
      } else {
        new Uint32Array(wireMsg.buffer, wireMsg.bufferOffset, 1)[0] = this.value.length

        const strLen = this.value.length
        const buf8 = new Uint8Array(wireMsg.buffer, wireMsg.bufferOffset + 4, strLen)
        for (let i = 0; i < strLen; i++) {
          buf8[i] = this.value[i].charCodeAt(0)
        }
      }
      wireMsg.bufferOffset += this.size
    }
  }
}

export function array(arg: ArrayBufferView): MessageMarshallingContext<ArrayBufferView, 'a', number> {
  return {
    value: arg,
    type: 'a',
    size: 4 + (function () {
      // fancy logic to calculate size with padding to a multiple of 4 bytes (int).
      return (arg.byteLength + 3) & ~3
    })(),
    optional: false,
    _marshallArg: function (wireMsg) {
      new Uint32Array(wireMsg.buffer, wireMsg.bufferOffset, 1)[0] = this.value.byteLength

      const byteLength = this.value.byteLength
      new Uint8Array(wireMsg.buffer, wireMsg.bufferOffset + 4, byteLength).set(new Uint8Array(this.value.buffer, 0, byteLength))

      wireMsg.bufferOffset += this.size
    }
  }
}

export function arrayOptional(arg: ArrayBufferView): MessageMarshallingContext<ArrayBufferView, 'a', number> {
  return {
    value: arg,
    type: 'a',
    size: 4 + (function () {
      if (arg === null) {
        return 0
      } else {
        // fancy logic to calculate size with padding to a multiple of 4 bytes (int).
        return (arg.byteLength + 3) & ~3
      }
    })(),
    optional: true,
    _marshallArg: function (wireMsg) {
      if (this.value === null) {
        new Uint32Array(wireMsg.buffer, wireMsg.bufferOffset, 1)[0] = 0
      } else {
        new Uint32Array(wireMsg.buffer, wireMsg.bufferOffset, 1)[0] = this.value.byteLength

        const byteLength = this.value.byteLength
        new Uint8Array(wireMsg.buffer, wireMsg.bufferOffset + 4, byteLength).set(new Uint8Array(this.value.buffer, 0, byteLength))
      }
      wireMsg.bufferOffset += this.size
    }
  }
}

function checkMessageSize(message: WlMessage, consumption: number) {
  if (message.consumed + consumption > message.size) {
    throw new Error(`Request too short.`)
  } else {
    message.consumed += consumption
  }
}

export function u(message: WlMessage): number {
  checkMessageSize(message, 4)
  return message.buffer[message.bufferOffset++]
}

export function i(message: WlMessage): number {
  checkMessageSize(message, 4)
  const arg = new Int32Array(message.buffer.buffer, message.buffer.byteOffset + (message.bufferOffset * Uint32Array.BYTES_PER_ELEMENT), 1)[0]
  message.bufferOffset += 1
  return arg
}

export function f(message: WlMessage): Fixed {
  checkMessageSize(message, 4)
  const arg = new Int32Array(message.buffer.buffer, message.buffer.byteOffset + (message.bufferOffset * Uint32Array.BYTES_PER_ELEMENT), 1)[0]
  message.bufferOffset += 1
  return new Fixed(arg >> 0)
}

export function o(message: WlMessage, optional: boolean, connection: Connection): WlObject | undefined {
  checkMessageSize(message, 4)
  const arg = message.buffer[message.bufferOffset++]
  if (optional && arg === 0) {
    return undefined
  } else {
    const wlObject = connection.wlObjects[arg]
    if (wlObject) {
      return wlObject
    } else {
      throw new Error(`Unknown object id ${arg}`)
    }
  }
}

export function n(message: WlMessage): number {
  checkMessageSize(message, 4)
  return message.buffer[message.bufferOffset++]
}

export function s(message: WlMessage, optional: boolean): string | undefined { // {String}
  checkMessageSize(message, 4)
  const stringSize = message.buffer[message.bufferOffset++]
  if (optional && stringSize === 0) {
    return undefined
  } else {
    const alignedSize = ((stringSize + 3) & ~3)
    checkMessageSize(message, alignedSize)
    // size -1 to eliminate null byte
    const byteArray = new Uint8Array(message.buffer.buffer, message.buffer.byteOffset + (message.bufferOffset * Uint32Array.BYTES_PER_ELEMENT), stringSize - 1)
    message.bufferOffset += (alignedSize / 4)
    return textDecoder.decode(byteArray)
  }
}

export function a(message: WlMessage, optional: boolean): ArrayBuffer | undefined {
  checkMessageSize(message, 4)
  const arraySize = message.buffer[message.bufferOffset++]
  if (optional && arraySize === 0) {
    return undefined
  } else {
    const alignedSize = ((arraySize + 3) & ~3)
    checkMessageSize(message, alignedSize)
    const arg = message.buffer.buffer.slice(message.buffer.byteOffset + (message.bufferOffset * Uint32Array.BYTES_PER_ELEMENT), message.buffer.byteOffset + (message.bufferOffset * Uint32Array.BYTES_PER_ELEMENT) + arraySize)
    message.bufferOffset += alignedSize
    return arg
  }
}

export function h(message: WlMessage): WebFD { // file descriptor {number}
  if (message.fds.length > 0) {
    let webFd = message.fds.shift();
    if (webFd === undefined) {
      throw new Error('No more webfds found in wl message.')
    }
    return webFd
  } else {
    throw new Error('Not enough file descriptors in message object.')
  }
}

export class Connection {
  readonly wlObjects: { [key: number]: WlObject } = {}
  closed: boolean = false
  onFlush?: (outMsg: SendMessage[]) => void
  private _outMessages: SendMessage[] = []
  private _inMessages: WlMessage[] = []
  private _idleHandlers: (() => void)[] = []

  /**
   * Adds a one-shot idle handler. The idle handler is fired once, after all incoming request messages have been processed.
   */
  addIdleHandler(idleHandler: () => void) {
    this._idleHandlers = [...this._idleHandlers, idleHandler]
  }

  removeIdleHandler(idleHandler: () => void) {
    this._idleHandlers = this._idleHandlers.filter(handler => handler !== idleHandler)
  }

  marshallMsg(id: number, opcode: number, size: number, argsArray: MessageMarshallingContext<any, any, any>[]) {
    const wireMsg = {
      buffer: new ArrayBuffer(size),
      fds: [],
      bufferOffset: 0
    }

    // write actual wire message
    const bufu32 = new Uint32Array(wireMsg.buffer)
    const bufu16 = new Uint16Array(wireMsg.buffer)
    bufu32[0] = id
    bufu16[2] = opcode
    bufu16[3] = size
    wireMsg.bufferOffset = 8

    // write actual argument value to buffer
    argsArray.forEach((arg) => arg._marshallArg(wireMsg))
    this.onSend(wireMsg)
  }

  private async _idle(): Promise<void> {
    for (const idleHandler of this._idleHandlers) {
      await idleHandler()
    }
  }

  /**
   * Handle received wire messages.
   */
  async message(incomingWireMessages: { buffer: Uint32Array, fds: Array<WebFD> }): Promise<void> {
    if (this.closed) {
      return
    }

    // more than one message in queue means the message loop is in await, don't concurrently process the new
    // message, instead return early and let the resume-from-await pick up the newly queued message.
    if (this._inMessages.push({
      ...incomingWireMessages, bufferOffset: 0,
      consumed: 0,
      size: 0
    }) > 1) {
      return
    }

    while (this._inMessages.length) {
      const wireMessages = this._inMessages[0]
      while (wireMessages.bufferOffset < wireMessages.buffer.length) {
        const id = wireMessages.buffer[wireMessages.bufferOffset]
        const sizeOpcode = wireMessages.buffer[wireMessages.bufferOffset + 1]
        wireMessages.size = sizeOpcode >>> 16
        const opcode = sizeOpcode & 0x0000FFFF

        if (wireMessages.size > wireMessages.buffer.byteLength) {
          throw new Error('Request buffer too small')
        }

        const wlObject = this.wlObjects[id]
        if (wlObject) {
          wireMessages.bufferOffset += 2
          wireMessages.consumed = 8
          try {
            // @ts-ignore
            await wlObject[opcode](wireMessages)
          } catch (e) {
            console.error(`
wlObject: ${wlObject.constructor.name}[${opcode}](..)
name: ${e.name} message: ${e.message} text: ${e.text}
error object stack:
${e.stack}
`)
            this.close()
            throw e
          }
          if (this.closed) {
            return
          }
        } else {
          throw new Error(`invalid object ${id}`)
        }
      }
      this._inMessages.shift()
    }

    this.flush()

    await this._idle()
  }

  /**
   * This doesn't actually send the message, but queues it so it can be send on flush.
   */
  onSend(wireMsg: SendMessage) {
    if (this.closed) {
      return
    }

    this._outMessages.push(wireMsg)
  }

  flush() {
    if (this.closed) {
      return
    }
    if (this._outMessages.length === 0) {
      return
    }

    this.onFlush?.(this._outMessages)
    this._outMessages = []
  }

  close() {
    if (this.closed) {
      return
    }

    // destroy resources in descending order
    Object.values(this.wlObjects).sort((a, b) => a.id - b.id).forEach((wlObject) => wlObject.destroy())
    this.closed = true
  }

  registerWlObject(wlObject: WlObject) {
    if (this.closed) {
      return
    }
    if (wlObject.id in this.wlObjects) {
      throw new Error(`Illegal object id: ${wlObject.id}. Already registered.`)
    }
    this.wlObjects[wlObject.id] = wlObject
  }

  unregisterWlObject(wlObject: WlObject) {
    if (this.closed) {
      return
    }
    delete this.wlObjects[wlObject.id]
  }
}
